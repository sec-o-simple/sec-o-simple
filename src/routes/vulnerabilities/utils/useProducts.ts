import { TProductTreeBranch } from '@/routes/products/types/tProductTreeBranch'
import useDocumentStore from '@/utils/useDocumentStore'
import { useMemo } from 'react'

export default function useProducts() {
  const productTree = useDocumentStore((state) => state.products)
  const productDetails = useMemo(
    () => Object.values(productTree)?.flatMap((ptb) => getProductDetails(ptb)),
    [productTree],
  )

  const getProductVersions = (productId: string) =>
    getProductVersionsOfBranch(Object.values(productTree), productId)

  return { productDetails, getProductVersions }
}

export type TProductDetails = {
  vendorName?: string
  productName: string
  productId: string
}

function getProductDetails(
  branch: TProductTreeBranch,
  vendorLabel?: string,
): TProductDetails[] {
  if (branch.category === 'vendor') {
    return branch.subBranches.flatMap((b) => getProductDetails(b, branch.name))
  } else if (branch.category === 'product_name') {
    return [
      {
        vendorName: vendorLabel,
        productName: branch.name,
        productId: branch.id,
      },
    ]
  } else {
    return []
  }
}

function getProductVersionsOfBranch(
  productTree: TProductTreeBranch[],
  productId: string,
): TProductTreeBranch[] | undefined {
  for (let branch of productTree) {
    if (branch.category === 'product_name' && branch.id === productId) {
      return branch.subBranches
    } else {
      const versions = getProductVersionsOfBranch(branch.subBranches, productId)
      if (versions) {
        return versions
      }
    }
  }
}
