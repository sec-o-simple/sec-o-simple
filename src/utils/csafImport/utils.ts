import { TProductTreeBranch } from '@/routes/products/types/tProductTreeBranch'

export function getParentPTB(
  id: string,
  startingBranches: TProductTreeBranch[],
  currentParent?: TProductTreeBranch,
): TProductTreeBranch | undefined {
  for (const branch of startingBranches) {
    if (startingBranches?.find((x) => x.id === id)) {
      return currentParent
    } else {
      const nestedSearch = getParentPTB(id, branch.subBranches, branch)
      if (nestedSearch) {
        return nestedSearch
      }
    }
  }
}
