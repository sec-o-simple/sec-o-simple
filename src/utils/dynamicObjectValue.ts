export type DynamicObjectValueKey<T, R = string> = keyof T | ((item: T) => R)

export function getDynamicObjectValue<T extends object, R = string>(
  obj: T,
  dynamicObjectValueKey: DynamicObjectValueKey<T, R>,
) {
  return typeof dynamicObjectValueKey === 'function'
    ? (dynamicObjectValueKey as (item: T) => R)(obj)
    : (obj[dynamicObjectValueKey] as R)
}
