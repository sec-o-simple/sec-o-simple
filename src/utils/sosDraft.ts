import { TDocumentInformation } from '@/routes/document-information/types/tDocumentInformation'
import {
  TProductFamily,
  TProductTreeBranch,
} from '@/routes/products/types/tProductTreeBranch'
import { TRelationship } from '@/routes/products/types/tRelationship'
import { TVulnerability } from '@/routes/vulnerabilities/types/tVulnerability'
import { getFilename } from './csafExport/helpers'
import { download } from './download'
import useDocumentStore, {
  TSOSDocumentType,
  sosDocumentTypes,
} from './useDocumentStore'

export type SOSDraft = {
  sosDocumentType: TSOSDocumentType
  documentInformation: TDocumentInformation
  products: TProductTreeBranch[]
  families: TProductFamily[]
  relationships: TRelationship[]
  vulnerabilities: TVulnerability[]
}

export function useSOSExport() {
  const documentStore = useDocumentStore()

  const exportSOSDocument = () => {
    download(
      `${getFilename(documentStore.documentInformation.id)}.sos.json`,
      JSON.stringify(documentStore, null, 2),
    )
  }

  return { exportSOSDocument }
}

export function useSOSImport() {
  const setSOSDocumentType = useDocumentStore(
    (state) => state.setSOSDocumentType,
  )
  const updateDocumentInformation = useDocumentStore(
    (state) => state.updateDocumentInformation,
  )
  const updateProducts = useDocumentStore((state) => state.updateProducts)
  const updateFamilies = useDocumentStore((state) => state.updateFamilies)
  const updateRelationships = useDocumentStore(
    (state) => state.updateRelationships,
  )
  const updateVulnerabilities = useDocumentStore(
    (state) => state.updateVulnerabilities,
  )

  const getSOSDraft = (jsonObject: object): SOSDraft | undefined => {
    if (
      !('sosDocumentType' in jsonObject) ||
      typeof jsonObject.sosDocumentType !== 'string' ||
      !sosDocumentTypes.includes(jsonObject.sosDocumentType as TSOSDocumentType)
    ) {
      return
    }

    if (
      !('documentInformation' in jsonObject) ||
      typeof jsonObject.documentInformation !== 'object'
    ) {
      return
    }

    if (
      !('products' in jsonObject) ||
      !(jsonObject.products instanceof Array)
    ) {
      return
    }

    if (
      !('families' in jsonObject) ||
      !(jsonObject.families instanceof Array)
    ) {
      return
    }

    if (
      !('relationships' in jsonObject) ||
      typeof jsonObject.relationships !== 'object'
    ) {
      return
    }

    if (
      !('vulnerabilities' in jsonObject) ||
      typeof jsonObject.vulnerabilities !== 'object'
    ) {
      return
    }

    return {
      sosDocumentType: jsonObject.sosDocumentType,
      documentInformation: jsonObject.documentInformation,
      products: jsonObject.products,
      families: jsonObject.families,
      relationships: jsonObject.relationships,
      vulnerabilities: jsonObject.vulnerabilities,
    } as SOSDraft
  }

  const isSOSDraft = (jsonObject: object): boolean => {
    return !!getSOSDraft(jsonObject)
  }

  const importSOSDocument = (jsonObject: object): boolean => {
    const sosDraft = getSOSDraft(jsonObject)
    if (sosDraft) {
      importSOSDraft(sosDraft)
      return true
    }
    return false
  }

  const importSOSDraft = (sosDraft: SOSDraft) => {
    setSOSDocumentType(sosDraft.sosDocumentType)
    updateDocumentInformation(sosDraft.documentInformation)
    updateProducts(sosDraft.products)
    updateFamilies(sosDraft.families)
    updateRelationships(sosDraft.relationships)
    updateVulnerabilities(sosDraft.vulnerabilities)
  }

  return { isSOSDraft, importSOSDocument, importSOSDraft }
}
