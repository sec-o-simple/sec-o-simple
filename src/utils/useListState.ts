import { useState } from 'react'
import {
  DynamicObjectValueKey,
  getDynamicObjectValue,
} from './dynamicObjectValue'

export type UseListStateProps<T, IdType> = {
  initialData?: T[]
  idKey?: DynamicObjectValueKey<T, IdType>
  /** Generator function to create a new element */
  generator?: T | (() => T)
  onRemove?: (entry: T) => void
}

export type ListState<T, IdType = string> = {
  data: T[]
  setData: React.Dispatch<React.SetStateAction<T[]>>
  updateDataEntry: (updatedEntry: T) => void
  removeDataEntry: (entry: T) => void
  addDataEntry: () => T | undefined
  getId: (entry: T) => IdType
}

export function useListState<T extends object, IdType = string>(
  options?: UseListStateProps<T, IdType>,
): ListState<T, IdType> {
  const { initialData = [], idKey = 'id' as keyof T, generator } = options ?? {}
  const [data, setData] = useState<T[]>(initialData)

  const updateDataEntry = (updatedEntry: T) => {
    setData(
      data.map((e) => (getId(e) === getId(updatedEntry) ? updatedEntry : e)),
    )
  }

  const removeDataEntry = (entry: T) => {
    setData([...data.filter((e) => getId(e) !== getId(entry))])
    options?.onRemove?.(entry)
  }

  /** adds a new data entry to the state
   *  returns the id of the new data entry */
  const addDataEntry = () => {
    const entry = typeof generator === 'function' ? generator() : generator
    if (entry) {
      setData([...data, entry])
      return entry
    }
  }

  const getId = (entry: T) => getDynamicObjectValue(entry, idKey)

  return {
    data,
    setData,
    updateDataEntry,
    removeDataEntry,
    getId,
    addDataEntry,
  }
}
