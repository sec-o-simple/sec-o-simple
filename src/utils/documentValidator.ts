import { TDocumentStore } from '@/utils/useDocumentStore'
import { createCSAFDocument } from './csafExport/csafExport'

import validate from '@secvisogram/csaf-validator-lib/validate.js'
import * as basic from '@secvisogram/csaf-validator-lib/basic.js'
import { ValidationMessage } from './useValidationStore'

const tests: unknown[] = [...Object.values(basic)]

export interface ValidationResult {
  isValid: boolean
  messages: ValidationMessage[]
}

export async function validateDocument(
  documentStore: TDocumentStore,
): Promise<ValidationResult> {
  try {
    const csafDocument = createCSAFDocument(documentStore)
    const result = await validate(tests, csafDocument)

    // Collect all messages with their severity
    const messages: ValidationMessage[] = []

    result.tests.forEach((test) => {
      test.errors.forEach((error) => {
        messages.push({
          path: error.instancePath,
          message: error.message,
          severity: 'error',
        })
      })

      test.warnings.forEach((warning) => {
        messages.push({
          path: warning.instancePath,
          message: warning.message,
          severity: 'warning',
        })
      })

      test.infos.forEach((info) => {
        messages.push({
          path: info.instancePath,
          message: info.message,
          severity: 'info',
        })
      })
    })

    return {
      isValid: result.isValid,
      messages: messages,
    }
  } catch (error) {
    return {
      isValid: false,
      messages: [],
    }
  }
}
