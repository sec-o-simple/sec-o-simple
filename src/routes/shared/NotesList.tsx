import ComponentList from '@/components/forms/ComponentList'
import HSplit from '@/components/forms/HSplit'
import { Input, Textarea } from '@/components/forms/Input'
import Select from '@/components/forms/Select'
import VSplit from '@/components/forms/VSplit'
import { checkReadOnly } from '@/utils/template'
import { ListState } from '@/utils/useListState'
import { Chip } from '@heroui/chip'
import { SelectItem } from '@heroui/select'
import { uid } from 'uid'

export const noteCategories = [
  'description',
  'details',
  'faq',
  'general',
  'legal_disclaimer',
  'other',
  'summary',
] as const
export type TNoteCategory = (typeof noteCategories)[number]

export type TNote = {
  id: string
  category: TNoteCategory
  content: string
  title: string
}

export const NoteGenerator = (): TNote => ({
  id: uid(),
  title: '',
  category: 'description',
  content: '',
})

export function NotesList({
  notesListState,
  csafPath,
  isTouched = false,
}: {
  notesListState: ListState<TNote>,
  csafPath: string,
  isTouched?: boolean,
}) {
  return (
    <ComponentList
      listState={notesListState}
      title="title"
      content={(note, index) => (
        <NoteForm
          note={note}
          csafPath={`${csafPath}/${index}`}
          isTouched={isTouched}
          onChange={notesListState.updateDataEntry}
        />
      )}
      startContent={(note) => <CategoryChip note={note} />}
    />
  )
}

function CategoryChip({ note }: { note: TNote }) {
  return (
    <Chip color="primary" variant="flat" radius="md" size="lg">
      {note.category}
    </Chip>
  )
}

function NoteForm({
  note,
  csafPath,
  onChange,
  isTouched = false,
}: {
  note: TNote
  csafPath: string
  onChange: (note: TNote) => void
  isTouched?: boolean
}) {
  return (
    <VSplit className="pt-4">
      <HSplit className="items-start">
        <Select
          label="Note category"
          csafPath={`${csafPath}/category`}
          isTouched={isTouched}
          selectedKeys={[note.category]}
          onSelectionChange={(selected) => {
            onChange({
              ...note,
              category: [...selected][0] as TNoteCategory,
            })
          }}
          isDisabled={checkReadOnly(note, 'category')}
        >
          {noteCategories.map((key) => (
            <SelectItem key={key}>{key}</SelectItem>
          ))}
        </Select>
        <Input
          label="Title"
          isTouched={isTouched}
          csafPath={`${csafPath}/title`}
          value={note.title}
          onValueChange={(newValue) => onChange({ ...note, title: newValue })}
          autoFocus={true}
          isDisabled={checkReadOnly(note, 'title')}
        />
      </HSplit>
      <Textarea
        label="Note Content"
        isTouched={isTouched}
        csafPath={`${csafPath}/text`}
        value={note.content}
        onValueChange={(newValue) => onChange({ ...note, content: newValue })}
        isDisabled={checkReadOnly(note, 'content')}
      />
    </VSplit>
  )
}
