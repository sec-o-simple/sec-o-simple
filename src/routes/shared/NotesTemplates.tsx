import { useConfigStore } from '@/utils/useConfigStore'
import { ListState } from '@/utils/useListState'
import { faAdd } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from '@heroui/button'
import {
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  useDisclosure,
} from '@heroui/modal'
import { Listbox, ListboxItem } from '@heroui/react'
import { useTranslation } from 'react-i18next'
import { uid } from 'uid'
import { TNote } from './NotesList'

export function NotesTemplates({
  notesListState,
  templatePath,
}: {
  notesListState: ListState<TNote>
  templatePath: string
}) {
  const { t } = useTranslation()
  const { isOpen, onOpen, onOpenChange } = useDisclosure()
  const templates = useConfigStore(
    (state) => state.config?.template?.[templatePath],
  )

  if (!templates) {
    return null
  }

  return (
    <>
      <div className="flex justify-end">
        <Button
          onPress={onOpen}
          variant="flat"
          color="primary"
          className="w-auto"
          startContent={<FontAwesomeIcon icon={faAdd} />}
        >
          {t('notesTemplates.addTemplateButton')}
        </Button>
      </div>
      <Modal isOpen={isOpen} size="xl" onOpenChange={onOpenChange}>
        <ModalContent>
          {(onClose) => (
            <>
              <ModalHeader>
                {t('notesTemplates.selectTemplateTitle')}
              </ModalHeader>
              <ModalBody>
                <Listbox>
                  {Object.entries(templates || {}).map(([key, template]) => (
                    <ListboxItem
                      key={key}
                      textValue={template.title}
                      description={
                        <p>
                          {template.content.length > 150
                            ? template.content.substring(0, 150) + '...'
                            : template.content}
                        </p>
                      }
                      onClick={() => {
                        notesListState.setData((prev) => [
                          ...prev,
                          {
                            id: uid(),
                            readonly: true,
                            deletable: true,
                            title: template.title,
                            category: template.category,
                            content: template.content,
                          },
                        ])
                        onClose()
                      }}
                    >
                      {template.title}
                    </ListboxItem>
                  ))}
                </Listbox>
              </ModalBody>

              <ModalFooter>
                <Button variant="light" color="primary" onPress={onClose}>
                  {t('common.close')}
                </Button>
              </ModalFooter>
            </>
          )}
        </ModalContent>
      </Modal>
    </>
  )
}
