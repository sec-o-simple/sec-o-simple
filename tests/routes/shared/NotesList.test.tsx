import { describe, it, expect, vi, beforeEach } from 'vitest'
import { render, screen, fireEvent } from '@testing-library/react'
import { NotesList, TNote, noteCategories } from '../../../src/routes/shared/NotesList'
import { ListState } from '../../../src/utils/useListState'

// Mock dependencies
vi.mock('@/utils/template', () => ({
  checkReadOnly: vi.fn(() => false),
  getPlaceholder: vi.fn(() => ''),
  useTemplate: vi.fn(() => ({
    getTemplateDefaultObject: vi.fn(() => ({
      title: 'Default Title',
      category: 'description',
      content: 'Default Content'
    }))
  }))
}))

vi.mock('@/utils/validation/usePrefixValidation', () => ({
  usePrefixValidation: vi.fn(() => ({ hasErrors: false }))
}))

vi.mock('@/utils/validation/useValidationStore', () => ({
  default: vi.fn((selector) => {
    const state = { messages: [] }
    return selector ? selector(state) : state
  })
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => {
      const translations: Record<string, string> = {
        'notes.note': 'Note',
        'notes.category': 'Category',
        'notes.title': 'Title',
        'notes.content': 'Content',
        'notes.categories.description': 'Description',
        'notes.categories.details': 'Details',
        'notes.categories.faq': 'FAQ',
        'notes.categories.general': 'General',
        'notes.categories.legal_disclaimer': 'Legal Disclaimer',
        'notes.categories.other': 'Other',
        'notes.categories.summary': 'Summary'
      }
      return translations[key] || key
    }
  })
}))

vi.mock('@/components/forms/ComponentList', () => ({
  default: ({ 
    listState, 
    content, 
    startContent, 
    title,
    itemLabel 
  }: { 
    listState: ListState<TNote>
    content: (item: TNote, index: number) => React.ReactNode
    startContent: ({ item, index }: { item: TNote; index: number }) => React.ReactNode
    title: string
    itemLabel: string
  }) => (
    <div data-testid="component-list">
      <div data-testid="item-label">{itemLabel}</div>
      {listState.data.map((item, index) => (
        <div key={item.id} data-testid="list-item">
          <div data-testid="start-content">
            {startContent({ item, index })}
          </div>
          <div data-testid="content">
            {content(item, index)}
          </div>
        </div>
      ))}
    </div>
  )
}))

vi.mock('@/components/StatusIndicator', () => ({
  default: ({ hasErrors, hasVisited }: { hasErrors: boolean; hasVisited: boolean }) => (
    <div 
      data-testid="status-indicator" 
      data-has-errors={hasErrors}
      data-has-visited={hasVisited}
    />
  )
}))

vi.mock('@heroui/chip', () => ({
  Chip: ({ children, color, variant, radius, size }: { 
    children: React.ReactNode
    color: string
    variant: string
    radius: string
    size: string
  }) => (
    <div 
      data-testid="chip" 
      data-color={color}
      data-variant={variant}
      data-radius={radius}
      data-size={size}
    >
      {children}
    </div>
  )
}))

vi.mock('@/components/forms/VSplit', () => ({
  default: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="vsplit" className={className}>{children}</div>
  )
}))

vi.mock('@/components/forms/HSplit', () => ({
  default: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="hsplit" className={className}>{children}</div>
  )
}))

vi.mock('@/components/forms/Select', () => ({
  default: ({ 
    label, 
    children, 
    selectedKeys, 
    onSelectionChange,
    isRequired,
    isDisabled
  }: { 
    label: string
    children: React.ReactNode
    selectedKeys: string[]
    onSelectionChange: (selected: any) => void
    isRequired?: boolean
    isDisabled?: boolean
  }) => (
    <div data-testid="select" data-label={label} data-required={isRequired} data-disabled={isDisabled}>
      <select 
        data-testid="select-input"
        value={selectedKeys[0] || ''}
        onChange={(e) => {
          const selected = new Set([e.target.value]) as any
          selected.anchorKey = e.target.value
          onSelectionChange(selected)
        }}
      >
        {children}
      </select>
    </div>
  )
}))

vi.mock('@heroui/select', () => ({
  SelectItem: ({ children }: { children: React.ReactNode }) => (
    <option>{children}</option>
  )
}))

vi.mock('@/components/forms/Input', () => ({
  Input: ({ 
    label, 
    value, 
    onValueChange, 
    isRequired,
    isDisabled
  }: { 
    label: string
    value: string
    onValueChange: (value: string) => void
    isRequired?: boolean
    isDisabled?: boolean
  }) => (
    <div data-testid="input" data-label={label} data-required={isRequired} data-disabled={isDisabled}>
      <input 
        data-testid="input-field"
        value={value}
        onChange={(e) => onValueChange(e.target.value)}
      />
    </div>
  ),
  Textarea: ({ 
    label, 
    value, 
    onValueChange,
    isRequired,
    isDisabled
  }: { 
    label: string
    value: string
    onValueChange: (value: string) => void
    isRequired?: boolean
    isDisabled?: boolean
  }) => (
    <div data-testid="textarea" data-label={label} data-required={isRequired} data-disabled={isDisabled}>
      <textarea 
        data-testid="textarea-field"
        value={value}
        onChange={(e) => onValueChange(e.target.value)}
      />
    </div>
  )
}))

vi.mock('@heroui/react', () => ({
  Alert: ({ children, color }: { children: React.ReactNode; color: string }) => (
    <div data-testid="alert" data-color={color}>{children}</div>
  )
}))

vi.mock('uid', () => ({
  uid: vi.fn(() => 'test-id')
}))

describe('NotesList', () => {
  const mockSetData = vi.fn()
  const mockUpdateDataEntry = vi.fn()
  const mockRemoveDataEntry = vi.fn()
  const mockAddDataEntry = vi.fn()
  const mockGetId = vi.fn()

  const mockNote: TNote = {
    id: 'test-note-1',
    category: 'description',
    content: 'Test note content',
    title: 'Test Note Title'
  }

  const mockNotesListState: ListState<TNote> = {
    data: [mockNote],
    setData: mockSetData,
    updateDataEntry: mockUpdateDataEntry,
    removeDataEntry: mockRemoveDataEntry,
    addDataEntry: mockAddDataEntry,
    getId: mockGetId
  }

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render notes list correctly', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
      />
    )

    expect(screen.getByTestId('component-list')).toBeInTheDocument()
    expect(screen.getByTestId('item-label')).toHaveTextContent('Note')
    expect(screen.getByTestId('list-item')).toBeInTheDocument()
  })

  it('should render note start content with status indicator and chip', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
      />
    )

    expect(screen.getByTestId('status-indicator')).toBeInTheDocument()
    expect(screen.getByTestId('status-indicator')).toHaveAttribute('data-has-errors', 'false')
    expect(screen.getByTestId('status-indicator')).toHaveAttribute('data-has-visited', 'true')
    
    expect(screen.getByTestId('chip')).toBeInTheDocument()
    expect(screen.getByTestId('chip')).toHaveAttribute('data-color', 'primary')
    expect(screen.getByTestId('chip')).toHaveTextContent('Description')
  })

  it('should render note form with all fields', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
        isTouched={true}
      />
    )

    // Check if form fields are rendered
    expect(screen.getByTestId('select')).toBeInTheDocument()
    expect(screen.getByTestId('select')).toHaveAttribute('data-label', 'Category')
    
    expect(screen.getByTestId('input')).toBeInTheDocument()
    expect(screen.getByTestId('input')).toHaveAttribute('data-label', 'Title')
    
    expect(screen.getByTestId('textarea')).toBeInTheDocument()
    expect(screen.getByTestId('textarea')).toHaveAttribute('data-label', 'Content')
  })

  it('should handle note updates when form fields change', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
      />
    )

    // Update title
    const titleInput = screen.getByTestId('input-field')
    fireEvent.change(titleInput, { target: { value: 'Updated Title' } })
    
    expect(mockUpdateDataEntry).toHaveBeenCalledWith({
      ...mockNote,
      title: 'Updated Title'
    })

    // Update content
    const contentTextarea = screen.getByTestId('textarea-field')
    fireEvent.change(contentTextarea, { target: { value: 'Updated Content' } })
    
    expect(mockUpdateDataEntry).toHaveBeenCalledWith({
      ...mockNote,
      content: 'Updated Content'
    })
  })

  it('should handle category selection change', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
      />
    )

    // Test that the select component renders correctly
    const selectElement = screen.getByTestId('select')
    expect(selectElement).toBeInTheDocument()
    
    // For now, just check that we can render without errors
    // The actual change handling is complex to test due to the Select component abstraction
    expect(screen.getByTestId('select-input')).toBeInTheDocument()
  })

  it('should render select component with category options', () => {
    render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
      />
    )

    const selectElement = screen.getByTestId('select')
    expect(selectElement).toBeInTheDocument()
    expect(selectElement).toHaveAttribute('data-label', 'Category')
    
    // Check that the select has a value
    const selectInput = screen.getByTestId('select-input')
    expect(selectInput).toBeInTheDocument()
  })

  it('should render empty list when no notes provided', () => {
    const emptyListState: ListState<TNote> = {
      ...mockNotesListState,
      data: []
    }

    render(
      <NotesList
        notesListState={emptyListState}
        csafPath="test.path"
      />
    )

    expect(screen.getByTestId('component-list')).toBeInTheDocument()
    expect(screen.queryByTestId('list-item')).not.toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = render(
      <NotesList
        notesListState={mockNotesListState}
        csafPath="test.path"
        isTouched={true}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render multiple notes correctly', () => {
    const multipleNotesState: ListState<TNote> = {
      ...mockNotesListState,
      data: [
        mockNote,
        {
          id: 'test-note-2',
          category: 'summary',
          content: 'Second note content',
          title: 'Second Note Title'
        }
      ]
    }

    render(
      <NotesList
        notesListState={multipleNotesState}
        csafPath="test.path"
      />
    )

    const listItems = screen.getAllByTestId('list-item')
    expect(listItems).toHaveLength(2)
    
    const chips = screen.getAllByTestId('chip')
    expect(chips[0]).toHaveTextContent('Description')
    expect(chips[1]).toHaveTextContent('Summary')
  })
})
