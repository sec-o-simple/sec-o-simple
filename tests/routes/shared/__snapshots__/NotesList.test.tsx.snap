// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotesList > should match snapshot 1`] = `
<div
  data-testid="component-list"
>
  <div
    data-testid="item-label"
  >
    Note
  </div>
  <div
    data-testid="list-item"
  >
    <div
      data-testid="start-content"
    >
      <div
        data-has-errors="false"
        data-has-visited="true"
        data-testid="status-indicator"
      />
      <div
        data-color="primary"
        data-radius="md"
        data-size="lg"
        data-testid="chip"
        data-variant="flat"
      >
        Description
      </div>
    </div>
    <div
      data-testid="content"
    >
      <div
        class="pt-4"
        data-testid="vsplit"
      >
        <div
          class="items-start"
          data-testid="hsplit"
        >
          <div
            data-disabled="false"
            data-label="Category"
            data-required="true"
            data-testid="select"
          >
            <select
              data-testid="select-input"
            >
              <option>
                Description
              </option>
              <option>
                Details
              </option>
              <option>
                FAQ
              </option>
              <option>
                General
              </option>
              <option>
                Legal Disclaimer
              </option>
              <option>
                Other
              </option>
              <option>
                Summary
              </option>
            </select>
          </div>
          <div
            data-disabled="false"
            data-label="Title"
            data-required="true"
            data-testid="input"
          >
            <input
              data-testid="input-field"
              value="Test Note Title"
            />
          </div>
        </div>
        <div
          data-disabled="false"
          data-label="Content"
          data-required="true"
          data-testid="textarea"
        >
          <textarea
            data-testid="textarea-field"
          >
            Test note content
          </textarea>
        </div>
      </div>
    </div>
  </div>
</div>
`;
